storage/gcloud/aio/storage/storage.py

```diff
@@ -57,6 +57,50 @@ def __init__(self, *, service_file: Optional[Union[str, io.IOBase]] = None,
     def get_bucket(self, bucket_name: str) -> Bucket:
         return Bucket(self, bucket_name)
 
+    async def copy(self, bucket: str, object_name: str,
+                   destination_bucket: str, *, new_name: str = None,
+                   headers: dict = None, params: dict = None,
+                   timeout: int = 10,
+                   session: Optional[Session] = None) -> bytes:
+        token = await self.token.get()
+
+        if not new_name:
+            new_name = object_name
+
+        # Using `rewriteTo` is preferred in part because it is able to make
+        # multiple calls to fully copy an object.
+        #
+        # https://cloud.google.com/storage/docs/json_api/v1/objects/rewrite
+        url = (f"{API_ROOT}/{bucket}/o/{quote(object_name, safe='')}/rewriteTo"
+               f"/b/{destination_bucket}/o/{quote(new_name, safe='')}")
+
+        headers = headers or {}
+        headers.update({
+            'Authorization': f'Bearer {token}',
+            'Content-Length': '0',
+            'Content-Type': '',
+        })
+
+        params = params or {}
+
+        s = AioSession(session) if session else self.session
+        resp = await s.post(url, headers=headers, params=params,
+                            timeout=timeout)
+
+        data: dict = await resp.json()
+
+        # When files are too large, multiple calls to `rewriteTo` need to be
+        # made. To refer to the same copy job, the `rewriteToken` from the
+        # previous return payload is used in subsequent `rewriteTo` calls. The
+        # user then has 1 week to complete the copy job.
+        while not data.get('done') and data.get('rewriteToken'):
```

Comment:
would it be useful to wrap this while in something like:

`try:
    while:
        retry code
except Exception as Ex:
    logger.error(f`Exception {Ex} interrupted copy operation')
    raise Ex`

for posterity

maybe include some stats if they are available from the API or generate your own stats like `rewrite attempts` My concern comes from the concurrency here as exceptions in other parts of the event loop would interrupt this code 

thoughts?

Reply:
I'll address each of your points:

1. The `rewriteToken` is not a retry operation, but a mechanic exposed by GCP to handle large copies. If the `rewriteTo` operation is incomplete, GCP requests that we keep following the API chain until it is complete. In that sense, it doesn't throw and we don't have to use a `try-except` block here.
2. To the best of my knowledge, exceptions in the event loop as a whole don't stop other coroutines from running unless the exception is not caught and crashes the program. In that case, the `rewriteTo` operation will expire in 7 days if not resumed and we're back to our initial state, so there's nothing to worry about. If the call is made again, then to the best of my knowledge a new `rewriteTo` operation is created and again after 7 days the previously unfinished one expired.

Does that change the context of your comments now?

Reply:
In addition, the user can always restart the copy if they have the `rewriteToken` by modifying the `params`. I can add that to the docstring.